### PR TITLE
SNOW-829219: Fix retry in LargeResultSet

### PR DESCRIPTION
**Pull request by sfc-gh-dprzybysz**
### Description
- Add tests for LargeResultSet handing timeout and 503 errors
- handle urllib timeout errors and not OK http response codes in LargeResultSet
- write err or body to callback after all retries in LargeResultSet

### Checklist
- [x] Format code according to the existing code style (there is no automatic linter yet)
- [x] Create tests which fail without the change (if possible)
- [x] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [x] Extend the README / documentation and ensure is properly displayed (if necessary)
- [x] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: lib/http/base.js [hunk 1]
```diff
@@ -57,11 +57,15 @@ HttpClient.prototype.request = function (options)
...
   {
-    var url = options.url;
+    const url = options.url;
 
-    Logger.getInstance().trace(url);
+    const timeout = options.timeout ||
+      this._connectionConfig.getTimeout() ||
+      DEFAULT_REQUEST_TIMEOUT;
 
+    Logger.getInstance().trace(`CALL ${options.method} with timout ${timeout}: ${url}`);
```
Comment (odedpeer) [inline lib/http/base.js:66]: typo

```suggestion
    Logger.getInstance().trace(`CALL ${options.method} with timeout ${timeout}: ${url}`);
```

File: lib/services/large_result_set.js [hunk 2]
```diff
@@ -20,13 +21,30 @@ function LargeResultSetService(connectionConfig, httpClient)
   Errors.assertInternal(Util.isObject(connectionConfig));
   Errors.assertInternal(Util.isObject(httpClient));
 
+  /**
+   * Does not HTTP client error should be retried
```
Comment (odedpeer) [inline lib/services/large_result_set.js:25]:
```suggestion
   * Should HTTP client error be retried
```

File: lib/http/base.js (hunk 1 above)
Comment (sfc-gh-dprzybysz) [reply lib/http/base.js]: Fixed, thanks

File: lib/services/large_result_set.js (hunk 2 above)
Comment (sfc-gh-dprzybysz) [reply lib/services/large_result_set.js]: Fixed, thanks

**Review by sfc-gh-pmotacki — APPROVED** [review]
LGTM

File: ci/container/hang_webserver.py [hunk 3]
```diff
@@ -20,6 +24,40 @@ def do_POST(self):
...
+        elif self.path.startswith('/xml'):
+            self.send_response(200, '<error/>')
+            self.send_header('Content-Type', 'application/xml')
+            self.end_headers()
+        elif self.path.startswith('/resetCounter'):
+            HTTPRequestHandler.counter = 0
+            self.send_response(200, 'OK')
+            self.send_header('Content-Type', 'text/plain')
+            self.end_headers()
+        elif self.path.startswith('/eachThirdReturns200Others503'):
```
Comment (sfc-gh-igarish) [inline ci/container/hang_webserver.py:40]: Could you please give details about "/eachThirdReturns200Others503"?

File: ci/container/hang_webserver.py [hunk 4]
```diff
@@ -20,6 +24,40 @@ def do_POST(self):
...
+            HTTPRequestHandler.counter += 1
+            if HTTPRequestHandler.counter % 3 == 0:
+                self.send_response(200, 'OK')
+                self.send_header('Content-Type', 'text/plain')
+                self.end_headers()
+            else:
+                self.send_response(503)
+                self.send_header('Content-Type', 'text/plain')
+                self.end_headers()
+        elif self.path.startswith('/eachThirdReturns200OthersHang'):
```
Comment (sfc-gh-igarish) [inline ci/container/hang_webserver.py:50]: Could you please give details about  /eachThirdReturns200OthersHang?

File: ci/container/hang_webserver.py (hunk 3 above)
Comment (sfc-gh-dprzybysz) [reply ci/container/hang_webserver.py:37]: I simplified hang webserver and added description for what do I need this endpoint

File: ci/container/hang_webserver.py (hunk 4 above)
Comment (sfc-gh-dprzybysz) [reply ci/container/hang_webserver.py:46]: I added description why I need this endpoint

**Review by sfc-gh-igarish — APPROVED** [review]
LGTM

**Pull request merged**